test/utils/fizzy_c_engine.cpp

```diff
@@ -0,0 +1,123 @@
+#include <cassert>
+#include <cstring>
+
+namespace fizzy::test
+{
+class FizzyCEngine final : public WasmEngine
+{
+    std::unique_ptr<FizzyInstance, void (*)(FizzyInstance*)> m_instance = {
+        nullptr, fizzy_free_instance};
+
+public:
+    bool parse(bytes_view input) const final;
+    std::optional<FuncRef> find_function(
+        std::string_view name, std::string_view signature) const final;
+    bool instantiate(bytes_view wasm_binary) final;
+    bool init_memory(bytes_view memory) final;
+    bytes_view get_memory() const final;
+    Result execute(FuncRef func_ref, const std::vector<uint64_t>& args) final;
+};
+
+namespace
+{
+FizzyExecutionResult env_adler32(void*, FizzyInstance* instance, const FizzyValue* args, int)
+{
+    auto* memory = fizzy_get_instance_memory_data(instance);
+    assert(memory != nullptr);
+    const auto size = fizzy_get_instance_memory_size(instance);
+    const auto ret = fizzy::test::adler32(
+        bytes_view(memory, size)
+            .substr(static_cast<uint32_t>(args[0].i64), static_cast<uint32_t>(args[1].i64)));
+    return {false, true, {ret}};
+}
+}  // namespace
+
+std::unique_ptr<WasmEngine> create_fizzy_c_engine()
+{
+    return std::make_unique<FizzyCEngine>();
+}
+
+bool FizzyCEngine::parse(bytes_view input) const
+{
+    const auto module = fizzy_parse(input.data(), input.size());
+    if (!module)
+        return false;
+
+    fizzy_free_module(module);
+    return true;
+}
+
+bool FizzyCEngine::instantiate(bytes_view wasm_binary)
+{
+    const auto module = fizzy_parse(wasm_binary.data(), wasm_binary.size());
+    if (!module)
+        return false;
+
+    FizzyValueType inputs[] = {FizzyValueTypeI32, FizzyValueTypeI32};
+    FizzyImportedFunction imports[] = {
+        {"env", "adler32", {{FizzyValueTypeI32, inputs, 2}, env_adler32, nullptr}}};
+    m_instance.reset(fizzy_resolve_instantiate(module, imports, 1));
+
+    return (m_instance != nullptr);
+}
+
+bool FizzyCEngine::init_memory(bytes_view memory)
+{
+    auto* instance_memory = fizzy_get_instance_memory_data(m_instance.get());
+    if (instance_memory == nullptr)
+        return false;
+
+    const auto size = fizzy_get_instance_memory_size(m_instance.get());
+    if (size < memory.size())
+        return false;
+
+    std::memcpy(instance_memory, memory.data(), memory.size());
+    return true;
+}
+
+bytes_view FizzyCEngine::get_memory() const
+{
+    auto* data = fizzy_get_instance_memory_data(m_instance.get());
+    if (data == nullptr)
+        return {};
+
+    const auto size = fizzy_get_instance_memory_size(m_instance.get());
+    return {data, size};
+}
+
+std::optional<WasmEngine::FuncRef> FizzyCEngine::find_function(
+    std::string_view name, std::string_view /*signature*/) const
+{
+    uint32_t func_idx;
+    if (!fizzy_find_exported_function(
+            fizzy_get_instance_module(m_instance.get()), std::string{name}.c_str(), &func_idx))
+    {
+        // TODO check function type
```

Comment:
Do we have this solved by the API?

Reply:
No, it's the same as in C++ API.

Reply:
`fizzy_engine.cpp` does check the signature.

Reply:
But i think we can ignore that because we decided to do it only once in fizzy_engine and ignore the other implementaitons.

Reply:
I thought to reuse the code from `fizzy_engine,cpp`, shoud I just remove TODO?

Reply:
Yeah, I think it is enough having it in one engine.